libV2/client/index.js

```diff
@@ -0,0 +1,248 @@
+                    keys: chunk.map(i => i.key),
+                });
+                callback();
+            },
+            error => {
+                this.push({
+                    success: false,
+                    keys: [],
+                });
+                moduleLogger.error('error uploading metrics from retry cache', { error });
+                callback();
+            });
+    }
+}
+
+class UtapiClient {
+    constructor(config) {
+        this._host = (config && config.host) || 'localhost';
+        this._port = (config && config.port) || '8100';
+        this._logger = (config && config.port) || moduleLogger;
+        this._maxCachedMetrics = (config && config.maxCachedMetrics) || 200000; // roughly 100MB
+        this._numCachedMetrics = 0;
+        this._retryCache = levelup(encode(memdown(), { valueEncoding: 'json' }));
+        this._drainTimer = null;
+        this._drainCanSchedule = true;
+        this._drainDelay = (config && config.drainDelay) || 30000;
+    }
+
+    async join() {
+        await this._flushRetryCacheToLogs();
+        this._retryCache.close();
+    }
+
+    async _pushToUtapi(metrics) {
+        const resp = await needle(
+            'post',
+            `http://${this._host}:${this._port}/v2/ingest`,
+            metrics.map(metric => metric.getValue()),
+            { json: true },
+        );
+        if (resp.statusCode !== 200) {
+            throw Error('failed to push metric, server returned non 200 status code',
+                { respCode: resp.statusCode, respMessage: resp.statusMessage });
+        }
+    }
+
+    async _addToRetryCache(metric) {
+        if (this._numCachedMetrics < this._maxCachedMetrics) {
+            try {
+                await this._retryCache.put(metric.uuid, metric.getValue());
+                this._numCachedMetrics += 1;
+                await this._scheduleDrain();
+                return true;
+            } catch (error) {
+                this._logger
+                    .error('error adding metric to retry cache', { error });
+                this._emitMetricLogLine(metric, { reason: 'error' });
+            }
+        } else {
+            this._emitMetricLogLine(metric, { reason: 'overflow' });
+        }
+        return false;
+    }
+
+    async _drainRetryCache() {
+        return new Promise((resolve, reject) => {
+            let success = true;
+            const toRemove = [];
+
+            this._retryCache.createReadStream()
+                .pipe(new Chunker({ chunksize: 100 }))
+                .pipe(new Uploader({ ingest: this._pushToUtapi.bind(this) }))
+                .on('data', res => {
+                    if (res.success) {
+                        toRemove.push(...res.keys);
+                    } else {
+                        success = false;
+                    }
+                })
+                .on('end', () => {
+                    this._retryCache.batch(toRemove.map(i => ({ type: 'del', key: i })));
```

Comment:
This call does not have an `await`, is it on purpose, like not to wait for this call to complete?